stan/math/prim/fun/matrix_exp_action_handler.hpp

```diff
@@ -104,26 +169,68 @@ class matrix_exp_action_handler {
    * @param [out] m int parameter m
    * @param [out] s int parameter s
    */
-  inline void set_approximation_parameter(const Eigen::MatrixXd& mat,
-                                          const double& t, int& m, int& s) {
-    if (l1norm(mat) < tol || t < tol) {
+  template <typename EigMat1, typename EigMat2,
+            require_all_eigen_t<EigMat1, EigMat2>* = nullptr,
+            require_all_st_same<double, EigMat1, EigMat2>* = nullptr>
+  inline void set_approx_order(const EigMat1& mat, const EigMat2& b,
+                               const double& t, int& m, int& s) {
+    // L1 norm
+    double normA = mat.colwise().template lpNorm<1>().maxCoeff();
+
+    if (normA < tol || t < tol) {
       m = 0;
       s = 1;
     } else {
-      m = m_max;
+      const std::vector<double>& theta = theta_m_double;
+      Eigen::VectorXd alpha(p_max - 1);
+
+      if (normA
+          < 4.0 * theta[m_max] * p_max * (p_max + 3) / (m_max * b.cols())) {
+        alpha = Eigen::VectorXd::Constant(p_max - 1, normA);
+      } else {
+        Eigen::VectorXd eta(p_max);
+        for (auto p = 0; p < p_max; ++p) {
+          double c = mat_power_1_norm(mat, p + 2);
+          c = std::pow(c, 1.0 / (p + 2.0));
+          eta[p] = c;
+        }
+        for (auto p = 0; p < p_max - 1; ++p) {
+          alpha[p] = std::max(eta[p], eta(p + 1));
+        }
+      }
+
+      Eigen::MatrixXd mt = Eigen::MatrixXd::Zero(p_max - 1, m_max);
+      for (auto p = 1; p < p_max; ++p) {
+        for (auto i = p * (p + 1) - 2; i < m_max; ++i) {
+          mt(p - 1, i) = alpha[p - 1] / theta[i];
+        }
+      }
 
-      Eigen::MatrixXd a = mat * t;
-      const double theta_m = theta_m_double_precision.back();
-      for (auto i = 0; i < std::ceil(std::log2(p_max)); ++i) {
-        a *= a;
+      Eigen::Matrix<int, -1, 1> u(m_max);
+      for (auto i = 0; i < m_max; ++i) {
+        u(i) = i + 1;
       }
-      double ap = std::pow(l1norm(a), 1.0 / p_max);
-      int c = std::ceil(ap / theta_m);
-      s = (c < 1 ? 1 : c);
+
+      Eigen::MatrixXd c = stan::math::ceil(mt) * u.asDiagonal();
+      for (auto i = 0; i < c.size(); ++i) {
+        if (c(i) <= 1.e-16) {
+          c(i) = std::numeric_limits<double>::infinity();
+        }
+      }
+
+      int cost = c.colwise().minCoeff().minCoeff(&m);
```

Comment:
Is this right? 

Do we need the first `.minCoeff()`? If so, why? It looks like we should be able to get by with the second one only?

Reply:
It is required. Compiler fails to compile. I'll leave it at that.